text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,296 @@
+We have to step back a bit: a breaking change is one that if it *didn't* come with an increasing major version, would result in valid Cargo plans that do not build.
+Yes, adding `empty-features = false` to a newer version of a crate will result in depending (not dependent) crate that do not specify a feature for that dependency being unable to upgrade.
+But that's it!
+Those crates will *not* have their dependencies upgraded incorrectly, resulting in a build failure, and thus there is no problem.
+
+(This is very much analogous to the situation with minimum Rust versions.
+Updating the minimum Rust version in a new version of a crate is not a breaking change.
+Some users with old versions will be unable to use the new version, but no one will be falsely upgraded resulting in a build failure.
+
+Indeed, that point of being able to express the minimum Rust version is precisely to *avoid* bumping the minimum supported version being a breaking change.
+Before, Cargo was unaware of this dependency on the compiler, and so raising the implicit lower bound had to be a breaking change.
+Now, Cargo is aware of the dependency and it is no longer a breaking change.
+
+Putting it altogether, in general we can say that bumping major versions is a tool of last result to account for issues Cargo can't directly know about.
+If Cargo can be directly made aware of the issue, then there is no "residual breakage" requiring a major version bump.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Bad names over time
+
+A downside of this proposal is that one has to resort to increasingly bad names.
+Take [seanmonstar/warp#968](https://github.com/seanmonstar/warp/issues/968) as an example.
+Warp currently ships HTTP2 support by default. This means that in order to be able to ship a version without http2, _all current features must depend on HTTP2_.
+This means creating new `websocket-without-http2`, `tls-without-http2`, `compression-without-http2`, `compression-brotli-without-http2`, `compression-gzip-without-http2` features in order for other crates to depend on them without pulling in http2 support.
+
+Note that this is only a problem for new features that existing features or orthogonal too.
+In many cases, existing features will naturally (not just for back compat) depend on the new features, and then there is no need or possibility to introduce `...-without-...` siblings to those features.
+
+My sense that while that is a downside, it is better to first walk that run.
+Do this and the problem is solved.
+We can add feature migrations later once there is more demand.
+See the Alternatives section for more details.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Empty feature set is empty library.
+
+An alternative is not to ban the empty feature set, but ensure it always translates to the empty library.
+I.e. to require that *all* items must be dependent upon *some* feature; everything needs a `cfg`.
```

Comment:
This is something a crate could already do, right? So in principle we don't need a new cargo feature to make this possible. The problem is that doing this to an existing crate requires one semver bump to make the no-feature-lib empty.

Reply:
Yes, right. I suppose making the `cargo init` crate have a `default` feature and `#![cfg(feature = "default")]` might do the trick.